File: javascript/pascals-triangle/pascals-triangle.js
function Triangle(amount) {
    this.rows = [[1]];
    this.lastRow = [1];
    for(var i=2; i<=amount; i++) {
        this.generateNextRow();
    }
}

Triangle.prototype.generateNextRow = function() {
    var nextRow = [];
    for(var i=0; i<=this.lastRow.length; i++){
        var left = this.lastRow[i-1] || 0;
        var right = this.lastRow[i] || 0;
        nextRow.push(left + right);
    }
    this.rows.push(nextRow);
    this.lastRow = nextRow;
}

module.exports = Triangle;

File: go/pascals-triangle/pascals_triangle.go
package pascal

const testVersion = 1

// Triangle creates the Pascal triangle with the given number of rows
func Triangle(amount int) [][]int {
	result := make([][]int, amount)
	for row := range result {
		result[row] = make([]int, row+1)
		result[row][0] = 1
		result[row][row] = 1
		for i := 1; i < row; i++ {
			result[row][i] = result[row-1][i-1] + result[row-1][i]
		}
	}
	return result
}
